sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableConfigTranslator.java

```diff
@@ -0,0 +1,342 @@
+      BatchingSettings tmpSettings = batchingSettings.build();
+      batchingSettings =
+          batchingSettings.setFlowControlSettings(
+              callSettings
+                  .getBatchingSettings()
+                  .getFlowControlSettings()
+                  .toBuilder()
+                  .setMaxOutstandingElementCount(
+                      tmpSettings.getElementCountThreshold() * writeOptions.getMaxRequests())
+                  .setMaxOutstandingRequestBytes(
+                      tmpSettings.getRequestByteThreshold() * writeOptions.getMaxRequests())
+                  .build());
+    }
+
+    settings
+        .stubSettings()
+        .bulkMutateRowsSettings()
+        .setRetrySettings(retrySettings.build())
+        .setBatchingSettings(batchingSettings.build());
+
+    return settings.build();
+  }
+
+  private static BigtableDataSettings configureReadSettings(
+      BigtableDataSettings.Builder settings, BigtableReadOptions readOptions) {
+
+    RetrySettings.Builder retrySettings =
+        settings.stubSettings().readRowsSettings().getRetrySettings().toBuilder();
+
+    if (readOptions.getAttemptTimeout() != null) {
+      retrySettings.setInitialRpcTimeout(Duration.ofMillis(readOptions.getAttemptTimeout()));
+
+      if (readOptions.getOperationTimeout() == null) {
+        retrySettings.setTotalTimeout(
+            Duration.ofMillis(
+                Math.max(
+                    retrySettings.getTotalTimeout().toMillis(), readOptions.getAttemptTimeout())));
+      }
+    }
+
+    if (readOptions.getOperationTimeout() != null) {
+      retrySettings.setTotalTimeout(Duration.ofMillis(readOptions.getOperationTimeout()));
+    }
+
+    if (readOptions.getRetryDelayMultiplier() != null) {
+      retrySettings.setRetryDelayMultiplier(readOptions.getRetryDelayMultiplier());
+    }
+
+    if (readOptions.getRetryInitialDelay() != null) {
+      retrySettings.setInitialRetryDelay(Duration.ofMillis(readOptions.getRetryInitialDelay()));
+    }
+
+    settings.stubSettings().readRowsSettings().setRetrySettings(retrySettings.build());
+
+    return settings.build();
+  }
+
+  /** Translate BigtableOptions to BigtableConfig. */
+  static BigtableConfig translateToBigtableConfig(BigtableConfig config, BigtableOptions options) {
+    BigtableConfig.Builder builder = config.toBuilder();
+
+    if (options.getProjectId() != null && config.getProjectId() == null) {
+      builder.setProjectId(ValueProvider.StaticValueProvider.of(options.getProjectId()));
+    }
+
+    if (options.getInstanceId() != null && config.getInstanceId() == null) {
+      builder.setInstanceId(ValueProvider.StaticValueProvider.of(options.getInstanceId()));
+    }
+
+    if (options.getAppProfileId() != null && config.getAppProfileId() == null) {
+      builder.setAppProfileId(ValueProvider.StaticValueProvider.of(options.getAppProfileId()));
+    }
+
+    if (!options.getDataHost().equals("bigtable.googleapis.com")
```

Comment:
what about batch? This would also break for regional endpoints

I think you might be better off checking if the options.get CredentialOptions().getCredentialType() == CredentialType.None